Type checker panics on bare `return` in unit-returning function

## Repro
\`\`\`mochi
fun noop(): unit {
  return
}
noop()
\`\`\`

\`mochi run\` panics:
\`\`\`
panic: runtime error: invalid memory address or nil pointer dereference
mochi/types.checkExprWithExpected(0x0, ...)
    types/check.go:1452 +0x28
mochi/types.checkStmt(...)
    types/check.go:1267
\`\`\`

## Cause
\`checkStmt\` for \`s.Return\` always calls \`checkExprWithExpected(s.Return.Value, env, expectedReturn)\`, but a bare \`return\` produces \`s.Return.Value == nil\`. \`checkExprWithExpected\` dereferences \`e.Binary\` unconditionally, hence the nil deref.

## Suggested fix
Treat a nil return value as \`UnitType{}\` and unify against the expected return. Snippet:

\`\`\`go
case s.Return != nil:
    if s.Return.Value == nil {
        if !unify(UnitType{}, expectedReturn, nil) {
            return errReturnMismatch(s.Return.Pos, expectedReturn, UnitType{})
        }
        return nil
    }
    // existing path...
\`\`\`

## Why this matters
Bare \`return\` is the natural early-exit shape for procedures, and \`unit\` just became a first-class surface type (#21225 / MEP-4 P19). The panic is observable from \`mochi run\` on any sample that combines them.

Discovered while writing the soundness fixture for P19 (#21244). Refs MEP-4 §6.

--- a/tests/types/soundness/p19b_bare_return_unit.expect
+++ b/tests/types/soundness/p19b_bare_return_unit.expect
@@ -1,0 +1,1 @@
+✅ Type Check Passed

--- a/tests/types/soundness/p19b_bare_return_unit.mochi
+++ b/tests/types/soundness/p19b_bare_return_unit.mochi
@@ -1,0 +1,7 @@
+// MEP-4 P19 follow-up (#21246): a bare `return` inside a unit-returning
+// function used to nil-deref the checker. The Return arm now treats a
+// missing value as `unit` and unifies it against the expected type.
+fun noop(): unit {
+  return
+}
+noop()

--- a/tests/types/soundness/p19c_bare_return_non_unit.error
+++ b/tests/types/soundness/p19c_bare_return_non_unit.error
@@ -1,0 +1,8 @@
+ 1. error[T010]: return type mismatch: expected int, got unit
+  --> tests/types/soundness/p19c_bare_return_non_unit.mochi:4:3
+
+  4 |   return
+    |   ^
+
+help:
+  Update the return value to match the function's return type.

--- a/tests/types/soundness/p19c_bare_return_non_unit.mochi
+++ b/tests/types/soundness/p19c_bare_return_non_unit.mochi
@@ -1,0 +1,5 @@
+// MEP-4 P19 follow-up: a bare `return` in a function that promised a
+// non-unit return type is a return-type mismatch, not a nil-deref panic.
+fun answer(): int {
+  return
+}

--- a/types/check.go
+++ b/types/check.go
@@ -1296,6 +1296,12 @@ func checkStmt(s *parser.Statement, env *Env, expectedReturn Type, inLoop bool) 
 		return err
 
 	case s.Return != nil:
+		if s.Return.Value == nil {
+			if !unify(UnitType{}, expectedReturn, nil) {
+				return errReturnMismatch(s.Return.Pos, expectedReturn, UnitType{})
+			}
+			return nil
+		}
 		actual, err := checkExprWithExpected(s.Return.Value, env, expectedReturn)
 		if err != nil {
 			return err

